blockchain/validate_test.go

```diff
@@ -313,66 +315,69 @@ func TestSigOpsLimitsWithMultiMBBlocks(t *testing.T) {
 	// tx1:      1 sigOps    <=1MB
 	// tx2:      1 sigOps    <=1MB
 	// block:    3 sigOps    < 1MB
-
-	// Add many NOOPs to the coinbase out script which can be arbitrary.
-	// This gets the block >1MB without breaking Tx size limits
-	bigEnoughScript := append(baseScriptForSize, repeatScript(979800, txscript.OP_NOP)...)
+	// Add 3 * 400k NOOPs to get >1MB block without breaking Tx size limits
+	manyNoOps := repeatScript(400000, txscript.OP_NOP)
+	for _, tx := range baseBlock.Transactions {
+		tx.TxOut[0].PkScript = append(tx.TxOut[0].PkScript, manyNoOps...)
+	}
 
 	// 1. Pass at per-Tx sigOps limit
 	// coinbase:     20k sigOps (ok)     <=1MB (ok)
 	// tx1:            1 sigOps (ok)     <=1MB (ok)
 	// tx2:            1 sigOps (ok)     <=1MB (ok)
 	// block:    20k + 2 sigOps (ok)     > 1MB (ok)
-	baseBlock.Transactions[0].TxOut[0].PkScript = bigEnoughScript
 	atTxLimit, err := newTestBlock(baseBlock, tip, 20000, 1, 1)
 	if err != nil {
 		t.Fatalf("Unexpected error creating sigOps test block: %v", err)
 	}
-	err = attachBigBlock(atTxLimit)
+	err = validateBigBlock(atTxLimit)
 	if err != nil {
-		t.Fatalf("Expected to validate but got: %v", err)
+		if rule, ok := err.(RuleError); ok {
+			if rule.ErrorCode == ErrTooManySigOps {
+				t.Fatalf("Expected to validate but got: %v", err)
+			}
+		}
+
 	}
 	tip = atTxLimit
 
 	// 1. Fail at per-Tx sigOps limit +1
-	// coinbase: 20k + 1 sigOps (fail)   <=1MB (ok)
+	// coinbase: 40k + 1 sigOps (fail)   <=1MB (ok)
 	// tx1:            1 sigOps (ok)     <=1MB (ok)
 	// tx2:            1 sigOps (ok)     <=1MB (ok)
-	// block:    20k + 3 sigOps (ok)     > 1MB (ok)
-	fmt.Println("1. Fail")
-	overTxLimit, err := newTestBlock(baseBlock, tip, 20001, 1, 1)
+	// block:    40k + 3 sigOps (ok)     > 1MB (ok)
+	overTxLimit, err := newTestBlock(baseBlock, tip, 40001, 1, 1)
 	if err != nil {
 		t.Fatalf("Unexpected error creating sigOps test block: %v", err)
 	}
-	err = attachBigBlock(overTxLimit)
-	isExpectedErr := false
+	err = validateBigBlock(overTxLimit)
 	if err != nil {
 		if rule, ok := err.(RuleError); ok {
-			if rule.ErrorCode == ErrTxTooManySigOps {
-				isExpectedErr = true
+			if rule.ErrorCode != ErrTooManySigOps {
+				t.Fatalf("Expected to fail with TooManySigOps but got: %v", err)
 			}
 		}
 	}
```

Comment:
This is silencing all other errors. Specifically it is silencing an error here:
`checkConnectBlock` --> `CheckTransactionInputs()`
which happens to occur just after the block-level sigops check. So it's checking block-level tx.
But (unless I am wrong!) it is not checking tx-level sigops which is what this part of the test is supposed to check.
In other words, even if we fix the error that is occuring, this test will fail because there is no way for it to raise the tx-level sigops error.

Reply:
What happened here is that you broke the block-level sigops test and got the block-level error by changing from 20k --> 40k.

Reply:
The only way I see to check tx-level sigops through regular block-attaching machinery is by using `ProcessBlock` which brings in several additional issues, especially POW.

Reply:
You switched to `ProcessBlock` so this should to tx level checks now. Also TIL about `BFNoPoWCheck`.